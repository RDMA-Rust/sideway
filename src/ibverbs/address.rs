use libc::IF_NAMESIZE;
use rdma_mummy_sys::{
    ibv_ah_attr, ibv_gid, ibv_gid_entry, ibv_global_route, IBV_GID_TYPE_IB, IBV_GID_TYPE_ROCE_V1, IBV_GID_TYPE_ROCE_V2,
};
use serde::{Deserialize, Serialize};
use std::ffi::CStr;
use std::io;
use std::{fmt, mem::MaybeUninit, net::Ipv6Addr};

#[derive(Default, Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Gid {
    pub raw: [u8; 16],
}

impl AsMut<ibv_gid> for Gid {
    fn as_mut(&mut self) -> &mut ibv_gid {
        unsafe { &mut *self.raw.as_mut_ptr().cast::<ibv_gid>() }
    }
}

impl From<ibv_gid> for Gid {
    fn from(gid: ibv_gid) -> Self {
        Self {
            raw: unsafe { gid.raw },
        }
    }
}

impl From<Gid> for ibv_gid {
    fn from(mut gid: Gid) -> Self {
        *gid.as_mut()
    }
}

impl From<Gid> for Ipv6Addr {
    fn from(gid: Gid) -> Self {
        Ipv6Addr::from(gid.raw)
    }
}

impl From<Ipv6Addr> for Gid {
    fn from(addr: Ipv6Addr) -> Self {
        Gid { raw: addr.octets() }
    }
}

impl fmt::Display for Gid {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for (i, &byte) in self.raw.iter().enumerate() {
            if i > 0 && i % 2 == 0 {
                write!(f, ":")?;
            }
            write!(f, "{byte:02x}")?;
        }
        Ok(())
    }
}

impl Gid {
    pub fn is_zero(&self) -> bool {
        let (prefix, aligned, suffix) = unsafe { self.raw.align_to::<u128>() };

        prefix.iter().all(|&x| x == 0) && suffix.iter().all(|&x| x == 0) && aligned.iter().all(|&x| x == 0)
    }

    pub fn is_unicast_link_local(&self) -> bool {
        self.raw[0] == 0xfe && self.raw[1] & 0xc0 == 0x80
    }
}

#[repr(u32)]
#[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
pub enum GidType {
    InfiniBand = IBV_GID_TYPE_IB,
    RoceV1 = IBV_GID_TYPE_ROCE_V1,
    RoceV2 = IBV_GID_TYPE_ROCE_V2,
}

impl From<u32> for GidType {
    fn from(gid_type: u32) -> Self {
        match gid_type {
            IBV_GID_TYPE_IB => GidType::InfiniBand,
            IBV_GID_TYPE_ROCE_V1 => GidType::RoceV1,
            IBV_GID_TYPE_ROCE_V2 => GidType::RoceV2,
            _ => panic!("Unknown Gid type: {gid_type}"),
        }
    }
}

#[repr(transparent)]
#[derive(Clone, Copy)]
pub struct GidEntry(pub(crate) ibv_gid_entry);

impl Default for GidEntry {
    fn default() -> Self {
        GidEntry(ibv_gid_entry {
            gid: Gid::default().into(),
            gid_index: 0,
            port_num: 1,
            gid_type: 0,
            ndev_ifindex: 0,
        })
    }
}

impl GidEntry {
    #[inline]
    pub fn gid_index(&self) -> u32 {
        self.0.gid_index
    }

    #[inline]
    pub fn port_num(&self) -> u32 {
        self.0.port_num
    }

    #[inline]
    pub fn gid_type(&self) -> GidType {
        self.0.gid_type.into()
    }

    #[inline]
    pub fn netdev_index(&self) -> u32 {
        self.0.ndev_ifindex
    }

    pub fn netdev_name(&self) -> Result<String, String> {
        let mut buf = vec![0u8; IF_NAMESIZE];

        let return_buf = unsafe { libc::if_indextoname(self.netdev_index(), buf.as_mut_ptr().cast()) };

        if return_buf.is_null() {
            return Err(format!("get netdev name failed {:?}", io::Error::last_os_error()));
        }

        Ok(CStr::from_bytes_until_nul(buf.as_slice())
            .unwrap()
            .to_str()
            .unwrap()
            .to_owned())
    }

    #[inline]
    pub fn gid(&self) -> Gid {
        unsafe { Gid { raw: self.0.gid.raw } }
    }
}

pub struct AddressHandleAttribute {
    pub(crate) attr: ibv_ah_attr,
}

impl Default for AddressHandleAttribute {
    fn default() -> Self {
        Self::new()
    }
}

impl AddressHandleAttribute {
    pub fn new() -> Self {
        AddressHandleAttribute {
            attr: unsafe { MaybeUninit::zeroed().assume_init() },
        }
    }

    pub fn setup_dest_lid(&mut self, dest_lid: u16) -> &mut Self {
        self.attr.dlid = dest_lid;
        self
    }

    pub fn setup_service_level(&mut self, sl: u8) -> &mut Self {
        self.attr.sl = sl;
        self
    }

    pub fn setup_port(&mut self, port_num: u8) -> &mut Self {
        self.attr.port_num = port_num;
        self
    }

    // TODO: should we setup the grh at once or set each fields separately?
    pub fn setup_grh(
        &mut self, dest_gid: &Gid, flow_label: u32, src_gid_index: u8, hop_limit: u8, traffic_class: u8,
    ) -> &mut Self {
        self.attr.grh = ibv_global_route {
            dgid: (*dest_gid).into(),
            flow_label,
            sgid_index: src_gid_index,
            hop_limit,
            traffic_class,
        };
        self.attr.is_global = 1;
        self
    }

    pub fn setup_grh_dest_gid(&mut self, dest_gid: &Gid) -> &mut Self {
        self.attr.grh.dgid = (*dest_gid).into();
        self.attr.is_global = 1;
        self
    }

    pub fn setup_grh_src_gid_index(&mut self, src_gid_index: u8) -> &mut Self {
        self.attr.grh.sgid_index = src_gid_index;
        self.attr.is_global = 1;
        self
    }

    pub fn setup_grh_hop_limit(&mut self, hop_limit: u8) -> &mut Self {
        self.attr.grh.hop_limit = hop_limit;
        self.attr.is_global = 1;
        self
    }
}

#[cfg(test)]
mod tests {
    use crate::ibverbs::address::Gid;
    use rdma_mummy_sys::ibv_gid;
    use rstest::rstest;
    use std::net::Ipv6Addr;
    use std::str::FromStr;

    #[rstest]
    #[case("fe80::", true)]
    #[case("fe80::1", true)]
    #[case("fd80::1", false)]
    #[case("2001:dead:beef:dead:beef:dead:beef:dead", false)]
    fn test_link_local_gid(#[case] ip_str: &str, #[case] expected: bool) {
        let ip = Ipv6Addr::from_str(ip_str).unwrap();

        let gid: Gid = ip.into();

        assert_eq!(expected, gid.is_unicast_link_local())
    }

    #[rstest]
    #[case([0xfe, 0x80, 0xde, 0xad, 0xbe, 0xef, 0xde, 0xad, 0xbe, 0xef, 0xde, 0xad, 0xbe, 0xef, 0xde, 0xad], "fe80:dead:beef:dead:beef:dead:beef:dead")]
    #[case([0xfe, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00], "fe80:0000:0000:0000:0000:0000:0000:0000")]
    fn test_from_ibv_gid(#[case] octets: [u8; 16], #[case] expected: &str) {
        let gid_ = ibv_gid { raw: octets };
        let gid = Gid::from(gid_);
        assert_eq!(format!("{gid}"), expected);
    }
}
