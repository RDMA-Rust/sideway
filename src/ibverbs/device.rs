//! The device is used for creating a device context, everything about RDMA starts here.
use std::ops::Index;
use std::{ffi::CStr, io, marker::PhantomData};

use rdma_mummy_sys::{
    ibv_device, ibv_free_device_list, ibv_get_device_guid, ibv_get_device_list, ibv_get_device_name, ibv_open_device,
    ibv_transport_type,
};

use super::device_context::DeviceContext;
use super::device_context::Guid;

/// Error returned by [`DeviceList::new`] for getting a new [`DeviceList`].
#[derive(Debug, thiserror::Error)]
#[error("failed to get device list")]
#[non_exhaustive]
pub struct GetDeviceListError(#[from] pub GetDeviceListErrorKind);

/// The enum type for [`GetDeviceListError`].
#[derive(Debug, thiserror::Error)]
#[error(transparent)]
#[non_exhaustive]
pub enum GetDeviceListErrorKind {
    Ibverbs(#[from] io::Error),
}

/// Error returned by [`Device::open`] for open the device to create a [`DeviceContext`].
#[derive(Debug, thiserror::Error)]
#[error("failed to open device")]
#[non_exhaustive]
pub struct OpenDeviceError(#[from] pub OpenDeviceErrorKind);

/// The enum type for [`OpenDeviceError`].
#[derive(Debug, thiserror::Error)]
#[error(transparent)]
#[non_exhaustive]
pub enum OpenDeviceErrorKind {
    Ibverbs(#[from] io::Error),
}

/// The RDMA device list which contains all RDMA devices based on the environment configuration.
pub struct DeviceList {
    devices: *mut *mut ibv_device,
    num_devices: usize,
}

impl DeviceList {
    /// Get a new RDMA device list based on current environment.
    pub fn new() -> Result<DeviceList, GetDeviceListError> {
        let mut num_devices: i32 = 0;
        let devices = unsafe { ibv_get_device_list(&mut num_devices as *mut _) };
        if devices.is_null() {
            return Err(GetDeviceListErrorKind::Ibverbs(io::Error::last_os_error()).into());
        }

        Ok(DeviceList {
            devices,
            num_devices: num_devices as usize,
        })
    }

    /// Get a device list iterator.
    pub fn iter(&self) -> DeviceListIter<'_> {
        DeviceListIter {
            current: 0,
            total: self.num_devices,
            devices: self,
        }
    }

    /// Get a device list slice from current device list.
    pub fn as_device_slice<'list>(&'list self) -> &'list [Device<'list>] {
        unsafe { std::slice::from_raw_parts(self.devices as *const Device<'list>, self.num_devices) }
    }

    /// Get the device from device list by index.
    pub fn get(&self, index: usize) -> Option<Device<'_>> {
        if index >= self.num_devices {
            None
        } else {
            unsafe {
                let device = *self.devices.add(index);
                if device.is_null() {
                    None
                } else {
                    Some(Device::new(device, self))
                }
            }
        }
    }

    /// Get the device list length.
    pub fn len(&self) -> usize {
        self.num_devices
    }

    /// Check if current device list is empty.
    pub fn is_empty(&self) -> bool {
        self.num_devices == 0
    }
}

impl<'list> Index<usize> for &'list DeviceList {
    type Output = Device<'list>;

    fn index(&self, index: usize) -> &Self::Output {
        &self.as_device_slice()[index]
    }
}

impl Drop for DeviceList {
    fn drop(&mut self) {
        unsafe { ibv_free_device_list(self.devices) };
    }
}

impl<'list> IntoIterator for &'list DeviceList {
    type Item = <DeviceListIter<'list> as Iterator>::Item;
    type IntoIter = DeviceListIter<'list>;

    fn into_iter(self) -> Self::IntoIter {
        DeviceListIter {
            current: 0,
            total: self.num_devices,
            devices: self,
        }
    }
}

/// The iterator of the [`DeviceList`].
pub struct DeviceListIter<'list> {
    current: usize,
    total: usize,
    devices: &'list DeviceList,
}

impl<'list> Iterator for DeviceListIter<'list> {
    type Item = Device<'list>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.current >= self.total {
            None
        } else {
            unsafe {
                let device = *self.devices.devices.add(self.current);
                self.current += 1;
                Some(Device::new(device, self.devices))
            }
        }
    }
}

/// The underlying transport type of the device.
#[repr(i32)]
#[derive(PartialEq, Eq, Debug)]
pub enum TransportType {
    Unknown = ibv_transport_type::IBV_TRANSPORT_UNKNOWN,
    InfiniBand = ibv_transport_type::IBV_TRANSPORT_IB,
    IWarp = ibv_transport_type::IBV_TRANSPORT_IWARP,
    Usnic = ibv_transport_type::IBV_TRANSPORT_USNIC,
    UsnicUdp = ibv_transport_type::IBV_TRANSPORT_USNIC_UDP,
    Unspecified = ibv_transport_type::IBV_TRANSPORT_UNSPECIFIED,
}

impl From<i32> for TransportType {
    fn from(trans: i32) -> Self {
        match trans {
            ibv_transport_type::IBV_TRANSPORT_UNKNOWN => TransportType::Unknown,
            ibv_transport_type::IBV_TRANSPORT_IB => TransportType::InfiniBand,
            ibv_transport_type::IBV_TRANSPORT_IWARP => TransportType::IWarp,
            ibv_transport_type::IBV_TRANSPORT_USNIC => TransportType::Usnic,
            ibv_transport_type::IBV_TRANSPORT_USNIC_UDP => TransportType::UsnicUdp,
            ibv_transport_type::IBV_TRANSPORT_UNSPECIFIED => TransportType::Unspecified,
            _ => panic!("Unknown transport type value: {trans}"),
        }
    }
}

impl std::fmt::Display for TransportType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TransportType::InfiniBand => write!(f, "InfiniBand"),
            TransportType::IWarp => write!(f, "iWARP"),
            TransportType::Usnic => write!(f, "usNIC"),
            TransportType::UsnicUdp => write!(f, "usNIC UDP"),
            TransportType::Unspecified => write!(f, "Unspecified"),
            TransportType::Unknown => write!(f, "Invalid transport"),
        }
    }
}

/// A safe wrapper around a raw RDMA device pointer.
///
/// The lifetime parameter ensures that a Device cannot outlive the DeviceList
/// from which it was derived.
#[derive(Clone, Copy, Debug)]
#[repr(transparent)]
pub struct Device<'list> {
    device: *mut ibv_device,
    _dev_list: PhantomData<&'list ()>,
}

impl Device<'_> {
    pub fn new(device: *mut ibv_device, _devices: &DeviceList) -> Self {
        Device {
            device,
            _dev_list: PhantomData,
        }
    }

    /// Open the device to create a [`DeviceContext`] for querying / creating all other RDMA
    /// resources later.
    pub fn open(&self) -> Result<DeviceContext, OpenDeviceError> {
        unsafe {
            let context = ibv_open_device(self.device);
            if context.is_null() {
                return Err(OpenDeviceErrorKind::Ibverbs(io::Error::last_os_error()).into());
            }
            Ok(DeviceContext { context })
        }
    }
}

/// Trait for common device information access.
pub trait DeviceInfo {
    /// Get the name of the device, for example, `mlx5_0`.
    fn name(&self) -> String;

    /// Get the GUID of the device.
    fn guid(&self) -> Guid;

    /// Get the transport type of the device.
    fn transport_type(&self) -> TransportType;
}

impl DeviceInfo for Device<'_> {
    fn name(&self) -> String {
        unsafe {
            let name = ibv_get_device_name(self.device);
            if name.is_null() {
                String::new()
            } else {
                String::from_utf8_lossy(CStr::from_ptr(name).to_bytes()).to_string()
            }
        }
    }

    fn guid(&self) -> Guid {
        unsafe { Guid(ibv_get_device_guid(self.device)) }
    }

    fn transport_type(&self) -> TransportType {
        unsafe { (*self.device).transport_type.into() }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_first_and_last() -> Result<(), Box<dyn std::error::Error>> {
        let devices = DeviceList::new().unwrap();

        assert!(devices.get(devices.len()).is_none());

        if !devices.is_empty() {
            let first = devices.get(0);
            let last = devices.get(devices.len() - 1);

            assert!(!first.unwrap().device.is_null());
            assert!(!last.unwrap().device.is_null());
        }

        Ok(())
    }

    #[test]
    #[should_panic(expected = "index out of bounds")]
    fn test_out_of_bound_index() {
        let devices = DeviceList::new().unwrap();

        let _ = (&devices)[devices.len()];
    }
}
