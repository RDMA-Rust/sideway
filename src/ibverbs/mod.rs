pub mod address;
pub mod completion;
pub mod device;
pub mod device_context;
pub mod memory_region;
pub mod protection_domain;
pub mod queue_pair;

use bitmask_enum::bitmask;
use rdma_mummy_sys::ibv_access_flags;

#[bitmask(i32)]
#[bitmask_config(vec_debug)]
pub enum AccessFlags {
    LocalWrite = ibv_access_flags::IBV_ACCESS_LOCAL_WRITE.0 as _,
    RemoteWrite = ibv_access_flags::IBV_ACCESS_REMOTE_WRITE.0 as _,
    RemoteRead = ibv_access_flags::IBV_ACCESS_REMOTE_READ.0 as _,
    RemoteAtomic = ibv_access_flags::IBV_ACCESS_REMOTE_ATOMIC.0 as _,
    MemoryWindowBind = ibv_access_flags::IBV_ACCESS_MW_BIND.0 as _,
    ZeroBased = ibv_access_flags::IBV_ACCESS_ZERO_BASED.0 as _,
    OnDemand = ibv_access_flags::IBV_ACCESS_ON_DEMAND.0 as _,
    HugeTlb = ibv_access_flags::IBV_ACCESS_HUGETLB.0 as _,
    FlushGlobal = ibv_access_flags::IBV_ACCESS_FLUSH_GLOBAL.0 as _,
    FlushPersistent = ibv_access_flags::IBV_ACCESS_FLUSH_PERSISTENT.0 as _,
    RelaxedOrdering = ibv_access_flags::IBV_ACCESS_RELAXED_ORDERING.0 as _,
}
