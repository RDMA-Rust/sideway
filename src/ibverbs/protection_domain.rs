//! A protection domain is used to associate [`QueuePair`]s with [`MemoryRegion`]s, as a means for
//! enabling and controlling network adapter access to Host System memory.
//!
//! [`QueuePair`]: crate::ibverbs::queue_pair::QueuePair
//!
use rdma_mummy_sys::{ibv_dealloc_pd, ibv_pd};
use std::marker::PhantomData;
use std::ptr::NonNull;

use super::{
    device_context::DeviceContext,
    memory_region::{MemoryRegion, RegisterMemoryRegionError},
    queue_pair::QueuePairBuilder,
    AccessFlags,
};

/// A protection domain that could be used to creating RDMA QP and RDMA MR on it to associate them
/// together.
#[derive(Debug)]
pub struct ProtectionDomain<'ctx> {
    pub(crate) pd: NonNull<ibv_pd>,
    _dev_ctx: PhantomData<&'ctx ()>,
}

unsafe impl Send for ProtectionDomain<'_> {}
unsafe impl Sync for ProtectionDomain<'_> {}

impl Drop for ProtectionDomain<'_> {
    fn drop(&mut self) {
        unsafe {
            ibv_dealloc_pd(self.pd.as_mut());
        }
    }
}

impl ProtectionDomain<'_> {
    pub(crate) fn new(_ctx: &DeviceContext, pd: NonNull<ibv_pd>) -> Self {
        ProtectionDomain {
            // This should not fail as the caller ensures pd is not NULL
            pd,
            _dev_ctx: PhantomData,
        }
    }

    /// Register a memory region that was allocated outside this module.
    ///
    /// # Safety
    ///
    /// The caller must ensure that `ptr` is valid for `len` bytes
    /// and that the memory remains accessible and unmodified as needed.
    pub unsafe fn reg_mr(
        &self, ptr: usize, len: usize, access: AccessFlags,
    ) -> Result<MemoryRegion, RegisterMemoryRegionError> {
        MemoryRegion::reg_mr(self, ptr, len, access)
    }

    /// Create a [`QueuePairBuilder`] for building QPs on this protection domain
    /// later.
    pub fn create_qp_builder(&self) -> QueuePairBuilder {
        QueuePairBuilder::new(self)
    }
}
