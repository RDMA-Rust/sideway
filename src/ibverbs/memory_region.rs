use rdma_mummy_sys::{ibv_dereg_mr, ibv_mr, ibv_reg_mr};
use std::{io, marker::PhantomData, ptr::NonNull};

use super::protection_domain::ProtectionDomain;
use super::AccessFlags;

#[derive(Debug, thiserror::Error)]
#[error("failed to register memory region")]
#[non_exhaustive]
pub struct RegisterMemoryRegionError(#[from] pub RegisterMemoryRegionErrorKind);

#[derive(Debug, thiserror::Error)]
#[error(transparent)]
#[non_exhaustive]
pub enum RegisterMemoryRegionErrorKind {
    Ibverbs(#[from] io::Error),
    #[error("unexpected null pointer provided")]
    UnexpectedNullPointer,
}

/// A registered memory region abstraction that wraps an RDMA memory region.
#[derive(Debug)]
pub struct MemoryRegion<'pd> {
    mr: NonNull<ibv_mr>,
    _pd: PhantomData<&'pd ()>,
}

impl Drop for MemoryRegion<'_> {
    fn drop(&mut self) {
        unsafe {
            ibv_dereg_mr(self.mr.as_mut());
        }
    }
}

impl MemoryRegion<'_> {
    /// Returns the RDMA local key.
    pub fn lkey(&self) -> u32 {
        unsafe { self.mr.as_ref().lkey }
    }

    /// Returns the RDMA remote key.
    pub fn rkey(&self) -> u32 {
        unsafe { self.mr.as_ref().rkey }
    }

    /// Returns the length of the registered region.
    pub fn region_len(&self) -> usize {
        unsafe { self.mr.as_ref().length }
    }

    /// Returns the starting pointer if the memory is host memory.
    /// Device memory is opaque and thus returns `None`.
    pub fn get_ptr(&self) -> usize {
        unsafe { self.mr.as_ref().addr as _ }
    }
}

impl<'pd> MemoryRegion<'pd> {
    /// Register a memory region that was allocated outside this module.
    ///
    /// # Safety
    ///
    /// The caller must ensure that `ptr` is valid for `len` bytes
    /// and that the memory remains accessible and unmodified as needed.
    pub(crate) unsafe fn reg_mr(
        pd: &'pd ProtectionDomain, ptr: usize, len: usize, access: AccessFlags,
    ) -> Result<Self, RegisterMemoryRegionError> {
        let mr = unsafe { ibv_reg_mr(pd.pd.as_ptr(), ptr as _, len, access.into()) };

        if mr.is_null() {
            return Err(RegisterMemoryRegionErrorKind::Ibverbs(io::Error::last_os_error()).into());
        }

        Ok(Self {
            mr: NonNull::new(mr).unwrap(),
            _pd: PhantomData,
        })
    }
}

// #[cfg(test)]
// mod tests {
//     use super::*;
//     use crate::ibverbs::{device, protection_domain::ProtectionDomain};
//     use std::ptr;

//     // Helper to get a PD for testing
//     fn get_test_pd() -> ProtectionDomain<'static> {
//         let devices = device::DeviceList::new().unwrap();
//         for device in &devices {
//             let ctx = device.open().unwrap();
//             return ctx.alloc_pd().unwrap()
//         }
//     }

//     #[test]
//     fn test_buffer_allocation() {
//         let buffer = Buffer::from_len_zeroed(1024);
//         assert_eq!(buffer.len, 1024);
//         assert!(!buffer.data.is_null());

//         // Verify buffer is zeroed
//         unsafe {
//             let slice = std::slice::from_raw_parts(buffer.data.as_ptr(), buffer.len);
//             assert!(slice.iter().all(|&x| x == 0));
//         }
//     }

//     #[test]
//     fn test_managed_mr_registration() {
//         let pd = get_test_pd();
//         let size = 4096;
//         let mr = pd.reg_managed_mr(size, AccessFlags::LocalWrite).unwrap();

//         assert_eq!(mr.region_len(), size);
//         assert_eq!(mr.memory_type(), MemoryType::Host);
//         assert!(mr.get_ptr().is_some());

//         // Verify memory properties
//         let ptr = mr.get_ptr().unwrap();
//         assert!(!ptr.is_null());
//         assert!(mr.lkey() != 0);
//         assert!(mr.rkey() != 0);
//     }

//     #[test]
//     fn test_user_mr_registration() {
//         let pd = get_test_pd();
//         let size = 4096;
//         let buffer = Buffer::from_len_zeroed(size);

//         unsafe {
//             let mr = pd.reg_user_mr(buffer.data.as_ptr(), size, AccessFlags::LocalWrite).unwrap();
//             assert_eq!(mr.region_len(), size);
//             assert_eq!(mr.memory_type(), MemoryType::Host);
//             assert_eq!(mr.get_ptr().unwrap(), buffer.data.as_ptr());
//         }
//     }

//     #[test]
//     fn test_device_mr_registration() {
//         let pd = get_test_pd();
//         let size = 4096;
//         let handle = DeviceMemoryHandle(0x1000); // Simulated device memory handle

//         unsafe {
//             let mr = pd.reg_device_mr(handle, size, AccessFlags::LocalWrite).unwrap();
//             assert_eq!(mr.region_len(), size);
//             assert_eq!(mr.memory_type(), MemoryType::Device);
//             assert_eq!(mr.handle(), handle.0);
//             assert!(mr.get_ptr().is_none()); // Device memory has no host pointer
//         }
//     }

//     #[test]
//     fn test_mr_error_handling() {
//         let pd = get_test_pd();

//         // Test null pointer error
//         unsafe {
//             let result = pd.reg_user_mr(ptr::null_mut(), 1024, AccessFlags::LocalWrite);
//             assert!(matches!(
//                 result.unwrap_err().0,
//                 RegisterMemoryRegionErrorKind::UnexpectedNullPointer
//             ));
//         }

//         // Test invalid device handle
//         unsafe {
//             let result = pd.reg_device_mr(DeviceMemoryHandle(0), 1024, AccessFlags::LocalWrite);
//             assert!(matches!(
//                 result.unwrap_err().0,
//                 RegisterMemoryRegionErrorKind::UnexpectedNullPointer
//             ));
//         }
//     }

//     #[test]
//     fn test_mr_offset_access() {
//         let pd = get_test_pd();
//         let size = 4096;
//         let mr = pd.reg_managed_mr(size, AccessFlags::LocalWrite).unwrap();

//         // Valid offset
//         assert!(mr.get_ptr_by_offset(100).is_some());

//         // Invalid offset
//         assert!(mr.get_ptr_by_offset(size + 1).is_none());
//     }

//     #[test]
//     fn test_memory_cleanup() {
//         let pd = get_test_pd();
//         let size = 4096;

//         // Test managed MR cleanup
//         {
//             let _mr = pd.reg_managed_mr(size, AccessFlags::LocalWrite).unwrap();
//             // MR should be deregistered and memory freed when dropped
//         }

//         // Test user MR cleanup
//         {
//             let buffer = Buffer::from_len_zeroed(size);
//             unsafe {
//                 let _mr = pd.reg_user_mr(buffer.data.as_ptr(), size, AccessFlags::LocalWrite).unwrap();
//                 // MR should be deregistered when dropped
//             }
//             // Buffer cleaned up here
//         }
//     }
// }
