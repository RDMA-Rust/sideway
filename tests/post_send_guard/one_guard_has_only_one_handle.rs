use sideway::ibverbs::{
    address::AddressHandleAttribute,
    device,
    device_context::Mtu,
    queue_pair::{PostSendGuard, QueuePair, QueuePairAttribute, QueuePairState, SetInlineData, WorkRequestFlags},
    AccessFlags,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let device_list = device::DeviceList::new()?;
    for device in &device_list {
        let ctx = device.open().unwrap();

        let pd = ctx.alloc_pd().unwrap();
        let data: Vec<u8> = vec![0; 64];
        let mr = unsafe {
            pd.reg_mr(
                data.as_ptr() as _,
                data.len(),
                AccessFlags::LocalWrite | AccessFlags::RemoteWrite,
            )
            .unwrap()
        };

        let _comp_channel = ctx.create_comp_channel().unwrap();
        let mut cq_builder = ctx.create_cq_builder();
        let sq = cq_builder.setup_cqe(128).build_ex().unwrap();
        let rq = cq_builder.setup_cqe(128).build_ex().unwrap();

        let mut builder = pd.create_qp_builder();

        // block for basic qp
        {
            let mut qp = builder
            .setup_max_inline_data(128)
            .setup_send_cq(&sq)
            .setup_recv_cq(&rq)
            .build()
            .unwrap();

            println!("qp pointer is {:?}", qp);
            // modify QP to INIT state
            let mut attr = QueuePairAttribute::new();
            attr.setup_state(QueuePairState::Init)
                .setup_pkey_index(0)
                .setup_port(1)
                .setup_access_flags(AccessFlags::LocalWrite | AccessFlags::RemoteWrite);
            qp.modify(&attr).unwrap();

            assert_eq!(QueuePairState::Init, qp.state());

            // modify QP to RTR state, set dest qp as itself
            let mut attr = QueuePairAttribute::new();
            attr.setup_state(QueuePairState::ReadyToReceive)
                .setup_path_mtu(Mtu::Mtu1024)
                .setup_dest_qp_num(qp.qp_number())
                .setup_rq_psn(1)
                .setup_max_dest_read_atomic(0)
                .setup_min_rnr_timer(0);
            // setup address vector
            let mut ah_attr = AddressHandleAttribute::new();
            let gid_entries = ctx.query_gid_table().unwrap();

            ah_attr
                .setup_dest_lid(1)
                .setup_port(1)
                .setup_service_level(1)
                .setup_grh_src_gid_index(gid_entries[0].gid_index().try_into().unwrap())
                .setup_grh_dest_gid(&gid_entries[0].gid())
                .setup_grh_hop_limit(64);
            attr.setup_address_vector(&ah_attr);
            qp.modify(&attr).unwrap();

            assert_eq!(QueuePairState::ReadyToReceive, qp.state());

            // modify QP to RTS state
            let mut attr = QueuePairAttribute::new();
            attr.setup_state(QueuePairState::ReadyToSend)
                .setup_sq_psn(1)
                .setup_timeout(12)
                .setup_retry_cnt(7)
                .setup_rnr_retry(7)
                .setup_max_read_atomic(0);

            qp.modify(&attr).unwrap();

            assert_eq!(QueuePairState::ReadyToSend, qp.state());

            let mut guard = qp.start_post_send();
            let buf = vec![0, 1, 2, 3];

            let write_handle = guard
                .construct_wr(233, WorkRequestFlags::Signaled | WorkRequestFlags::Inline)
                .setup_write(mr.rkey(), mr.get_ptr() as _);

            // while holding a write handle, we can't build a send handle at the same time
            let _send_handle = guard.construct_wr(2, 0.into()).setup_send();

            write_handle.setup_inline_data(&buf);
        }

        // block for extended qp
        {
            let mut qp = builder
            .setup_max_inline_data(128)
            .setup_send_cq(&sq)
            .setup_recv_cq(&rq)
            .build_ex()
            .unwrap();

            println!("qp pointer is {:?}", qp);
            // modify QP to INIT state
            let mut attr = QueuePairAttribute::new();
            attr.setup_state(QueuePairState::Init)
                .setup_pkey_index(0)
                .setup_port(1)
                .setup_access_flags(AccessFlags::LocalWrite | AccessFlags::RemoteWrite);
            qp.modify(&attr).unwrap();

            assert_eq!(QueuePairState::Init, qp.state());

            // modify QP to RTR state, set dest qp as itself
            let mut attr = QueuePairAttribute::new();
            attr.setup_state(QueuePairState::ReadyToReceive)
                .setup_path_mtu(Mtu::Mtu1024)
                .setup_dest_qp_num(qp.qp_number())
                .setup_rq_psn(1)
                .setup_max_dest_read_atomic(0)
                .setup_min_rnr_timer(0);
            // setup address vector
            let mut ah_attr = AddressHandleAttribute::new();
            let gid_entries = ctx.query_gid_table().unwrap();

            ah_attr
                .setup_dest_lid(1)
                .setup_port(1)
                .setup_service_level(1)
                .setup_grh_src_gid_index(gid_entries[0].gid_index().try_into().unwrap())
                .setup_grh_dest_gid(&gid_entries[0].gid())
                .setup_grh_hop_limit(64);
            attr.setup_address_vector(&ah_attr);
            qp.modify(&attr).unwrap();

            assert_eq!(QueuePairState::ReadyToReceive, qp.state());

            // modify QP to RTS state
            let mut attr = QueuePairAttribute::new();
            attr.setup_state(QueuePairState::ReadyToSend)
                .setup_sq_psn(1)
                .setup_timeout(12)
                .setup_retry_cnt(7)
                .setup_rnr_retry(7)
                .setup_max_read_atomic(0);

            qp.modify(&attr).unwrap();

            assert_eq!(QueuePairState::ReadyToSend, qp.state());

            let mut guard = qp.start_post_send();
            let buf = vec![0, 1, 2, 3];

            let write_handle = guard
                .construct_wr(233, WorkRequestFlags::Signaled)
                .setup_write(mr.rkey(), mr.get_ptr() as _);

            // while holding a write handle, we can't build a send handle at the same time
            let _send_handle = guard.construct_wr(2, 0.into()).setup_send();

            write_handle.setup_inline_data(&buf);
        }
    }

    Ok(())
}
