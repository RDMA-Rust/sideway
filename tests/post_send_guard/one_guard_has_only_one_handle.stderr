error[E0499]: cannot borrow `guard` as mutable more than once at a time
  --> tests/post_send_guard/one_guard_has_only_one_handle.rs:89:32
   |
84 |             let write_handle = guard
   |                                ----- first mutable borrow occurs here
...
89 |             let _send_handle = guard.construct_wr(2, 0.into()).setup_send();
   |                                ^^^^^ second mutable borrow occurs here
90 |
91 |             write_handle.setup_inline_data(&buf);
   |             ------------ first borrow later used here

error[E0499]: cannot borrow `guard` as mutable more than once at a time
   --> tests/post_send_guard/one_guard_has_only_one_handle.rs:159:32
    |
154 |             let write_handle = guard
    |                                ----- first mutable borrow occurs here
...
159 |             let _send_handle = guard.construct_wr(2, 0.into()).setup_send();
    |                                ^^^^^ second mutable borrow occurs here
160 |
161 |             write_handle.setup_inline_data(&buf);
    |             ------------ first borrow later used here
